google-cloud-firestore/src/test/java/com/google/cloud/firestore/FirestoreBundleTest.java

```diff
@@ -0,0 +1,325 @@
+            .putAllFields(equivalentSnapshot.getProtoFields())
+            .setCreateTime(equivalentSnapshot.getCreateTime().toProto())
+            .setUpdateTime(equivalentSnapshot.getUpdateTime().toProto())
+            .setName(expectedDocumentName)
+            .build(),
+        document);
+  }
+
+  public static void verifyNamedQuery(
+      NamedQuery namedQuery, String name, Timestamp readTime, Query query, LimitType limitType) {
+    assertEquals(
+        NamedQuery.newBuilder()
+            .setName(name)
+            .setReadTime(readTime)
+            .setBundledQuery(
+                BundledQuery.newBuilder()
+                    .setParent(query.toProto().getParent())
+                    .setStructuredQuery(query.toProto().getStructuredQuery())
+                    .setLimitType(limitType)
+                    .build())
+            .build(),
+        namedQuery);
+  }
+
+  @Test
+  public void bundleToElementListWorks() {
+    String bundleString =
+        "20{\"a\":\"string value\"}9{\"b\":123}26{\"c\":{\"d\":\"nested value\"}}";
+    List<String> elements =
+        bundleToElementList(ByteBuffer.wrap(bundleString.getBytes(StandardCharsets.UTF_8)));
+    assertArrayEquals(
+        new String[] {
+          "{\"a\":\"string value\"}", "{\"b\":123}", "{\"c\":{\"d\":\"nested value\"}}"
+        },
+        elements.toArray());
+  }
+
+  public static List<BundleElement> toBundleElements(ByteBuffer bundleBuffer) {
+    ArrayList<BundleElement> result = new ArrayList<>();
+    for (String s : bundleToElementList(bundleBuffer)) {
+      BundleElement.Builder b = BundleElement.newBuilder();
+      try {
+        parser.merge(s, b);
+      } catch (InvalidProtocolBufferException e) {
+        throw new RuntimeException(e);
+      }
+      result.add(b.build());
+    }
+
+    return result;
+  }
+
+  @Test
+  public void bundleWithDocumentSnapshots() {
+    FirestoreBundle.Builder bundleBuilder = new FirestoreBundle.Builder(TEST_BUNDLE_ID);
+    bundleBuilder.add(UPDATED_SINGLE_FIELD_SNAPSHOT);
+    bundleBuilder.add(SINGLE_FIELD_SNAPSHOT);
+    ByteBuffer bundleBuffer = bundleBuilder.build().toByteBuffer();
+
+    // Expected bundle elements are [bundleMetadata, meta of UPDATED_SINGLE_FIELD_SNAPSHOT,
+    // UPDATED_SINGLE_FIELD_SNAPSHOT]
+    // because UPDATED_SINGLE_FIELD_SNAPSHOT is newer.
+    List<BundleElement> elements = toBundleElements(bundleBuffer);
+    assertEquals(3, elements.size());
+
+    verifyMetadata(
+        elements.get(0).getMetadata(),
+        // Even this snapshot is not included, its read time is still used as create time
+        // because it is the largest read time came across.
+        UPDATED_SINGLE_FIELD_SNAPSHOT.getReadTime().toProto(),
+        /*totalDocuments*/ 1,
+        /*expectEmptyContent*/ false);
+
+    verifyDocumentAndMeta(
+        elements.get(1).getDocumentMetadata(),
+        elements.get(2).getDocument(),
+        DOCUMENT_NAME,
+        Lists.newArrayList(),
```

Comment:
The expected type parameter isn't carried through here (the type of the list is derived from the elements passed to the method), we have to specify it so the compiler is happy.
```suggestion
        Lists.<String>newArrayList(),
```

Reply:
Done.